File: Netlists/MIXED_SIGNAL/Python/ADCStateTest.py
# This netlist is used to test the getTimeStatePairsADC() 
# method for valid YADC names.  It also tests that the width 
# value on the instance line takes precedence over the width 
# value in the model card, as set by the device models.  It
# also tests the setting of the quantization levels works
# correctly in the constructor for the YADC device.

import sys
# file will be used to compare timeArray and stateArray against
# a gold standard
f= open("ADCStateTest.cir.TSarrayData","w+")

from xyce_interface import xyce_interface

# this calls the xyce_interface.open() method to
# make a xyce object

libDirectory = sys.argv[1]
xyceObj = xyce_interface(libdir=libDirectory)
print( xyceObj )

argv= ['ADCStateTest.cir']
print( "calling initialize with netlist %s" % argv[0] )

result = xyceObj.initialize(argv)
print( "return value from initialize is %d" % result )

# get ADC names;
(result, names) = xyceObj.getDeviceNames("YADC")
print( "return value from getDeviceNames is %d" % result )
print( names )

# get the ADC widths.  This is hard-coded for two ADCs.  The ordering in
# widths seems reversed but names came back as:
#   ['YADC!ADC2', 'YADC!ADC1']
widths=[3,2]
(result,widths) = xyceObj.getADCWidths(names)
print( "return value from getADCWidths is %d" % result )
print( widths )

stepSize = 1e-5
steps = range(0,10)
for i in steps:
  requested_time = 0.0 + (i+1) * stepSize
  print( "Calling simulateUntil for requested_time = %.5f" % requested_time )
  actual_time = 0.0
  (result, actual_time) = xyceObj.simulateUntil( requested_time )
  print( "simulateUntil status = %d and actual_time = %.5f" % (result, actual_time) )
  (result, ADCnames, numADCnames, numPoints, timeArray, stateArray) = xyceObj.getTimeStatePairsADC()
  print( "return value from getTimeStatePairsADC is %d" % result )
  print( "number of ADC names returned by getTimeStatePairsADC is %d" % numADCnames )
  print( ADCnames )
  # output to stdout (for human readability)
  print( "number of points returned by getTimeStatePairsADC is %d" % numPoints )
  # numPoints is the maximum number of time points for an ADC.  Others may have fewer points
  # so for each time array returned for a given ADC search for the current time (within some error)
  for adcnum in range(0, numADCnames ):
    prevPoint = -1;
    currPoint = 0;
    for j in range(0, numPoints):
      if (abs(timeArray[adcnum][j] - requested_time) < 1.0e-11):
        currPoint=j
        prevPoint=j-1
        break
    
    print( "ADC %d: Time and state array %d values are %.3e %d" %((adcnum+1), prevPoint, timeArray[adcnum][prevPoint] , stateArray[adcnum][prevPoint]) )
    print( "ADC %d: Time and state array %d values are %.3e %d" %((adcnum+1), currPoint, timeArray[adcnum][currPoint] , stateArray[adcnum][currPoint]) )
    # output to file (for comparison against a gold standard)
    f.write( "ADC %d: Time and state array %d values are %.3e %d\n" %((adcnum+1), prevPoint, timeArray[adcnum][prevPoint] , stateArray[adcnum][prevPoint]) )
    f.write( "ADC %d: Time and state array %d values are %.3e %d\n" %((adcnum+1), currPoint, timeArray[adcnum][currPoint] , stateArray[adcnum][currPoint]) )

print( "calling close")
xyceObj.close()

# close file also
f.close()

File: Netlists/MIXED_SIGNAL/VPI/ADCStateTest.c
#include <vpi_user.h>
#include <stdio.h>
#include <stdlib.h>
#include <N_CIR_XyceCInterface.h>

static int ADCStateTest_compiletf(char*user_data)
{
      return 0;
}

static int ADCStateTest_calltf(char*user_data)
{
      printf("In calltf for ADCStateTest\n");

      // Used as a pointer to a pointer to an N_CIR_Xyce object.
      // This somewhat convoluted syntax is needed to stop p from
      // pointing at the same address as the VPI system task.
      void** p = (void **) malloc( sizeof(void* [1]) );

      // Turn the desired Xyce command line invocation into an int and
      // char** pointer that can used to initialize an N_CIR_ Xyce object.
      // This is hard-coded for initial testing purposes.
      char *argList[] = {
	  (char*)("Xyce"),
          (char*)("-quiet"),
          (char*)("ADCStateTest.cir")
      };
      int argc = sizeof(argList)/sizeof(argList[0]);
      char** argv = argList;

      xyce_open(p);
      xyce_initialize(p,argc,argv);

      // Get the number of YADC in the netlist, and the maximum name length
      // This allows for better sizing of the ADCnames char arrays.
      int numADCnames, maxADCnameLength;
      int* numADCnamesPtr = &numADCnames;
      int* maxADCnameLengthPtr = &maxADCnameLength;
      xyce_getNumDevices(p, (char *)"YADC", numADCnamesPtr, maxADCnameLengthPtr);
      printf("Num ADCs and Max ADC Name Length: %d %d\n",numADCnames,maxADCnameLength);

      // Initialize arrays of char array
      char ** ADCnames;

      int i,j,k; // loop counters
      ADCnames = (char **) malloc( numADCnames * sizeof(char*));
      for (i = 0; i < numADCnames; i++)
      {
        ADCnames[i] = (char *) malloc( (maxADCnameLength+1)*sizeof(char) );
      }

      int status, numPoints;
      int* numPointsPtr = &numPoints;
      double stepSize = 1e-5;
      double requested_time, actual_time;

      double* actual_time_ptr = &actual_time;
      double** timeArray;
      int** stateArray;
      int maxNumberOfTimeStateVals = 25;

      timeArray = (double **) malloc(numADCnames * sizeof(double*));
      stateArray = (int **) malloc(numADCnames * sizeof(int*));
      for (i=0; i<numADCnames; i++)
      {
        timeArray[i] = (double *) malloc(maxNumberOfTimeStateVals*sizeof(double));
        stateArray[i] = (int *) malloc(maxNumberOfTimeStateVals*sizeof(int));
      }

      FILE *fptr;
      fptr = fopen("ADCStateTest.cir.TSarrayData","w" );
      
      for (i=0; i<10; i++)
      {
        
        requested_time = 0.0 + (i+1) * stepSize;
        printf( "Calling simulateUntil for requested time %f\n", requested_time );
        status = xyce_simulateUntil(p, requested_time, actual_time_ptr );
        printf( "Return status from simulateUntil = %d and actual_time = %f\n",status, actual_time);
        
        status = xyce_getTimeStatePairsADC(p, numADCnamesPtr, ADCnames, numPointsPtr, timeArray, stateArray);
        
        printf( "number of points returned by getTimeVoltagePairsADC is %d\n", numPoints );
        // output to stdout (for human readability)
          
        int adcNum;
        for ( adcNum=0; adcNum<numADCnames; adcNum++ )
        {
          int prev=0;
          int curr=0;
          for( j=0; j<numPoints; j++ )
          {
            if( abs( timeArray[adcNum][j] - actual_time ) < 1.0e-11 )
            {
              // found end time of this ADC's time array
              curr = j;
              prev = j-1;
              if( prev < 0 )
              {
                prev = 0;
              }
              break; 
            }
          } 
          // output to std out for inspection 
          printf( "ADC %d: Time and state array %d values are %.3e %d\n", (adcNum+1), prev, timeArray[adcNum][prev], stateArray[adcNum][prev] );
          printf( "ADC %d: Time and state array %d values are %.3e %d\n", (adcNum+1), curr, timeArray[adcNum][curr], stateArray[adcNum][curr] );

          // output to file (for comparison against a gold standard)
          fprintf( fptr, "ADC %d: Time and state array %d values are %.3e %d\n", (adcNum+1), prev, timeArray[adcNum][prev], stateArray[adcNum][prev] );
          fprintf( fptr, "ADC %d: Time and state array %d values are %.3e %d\n", (adcNum+1), curr, timeArray[adcNum][curr], stateArray[adcNum][curr] );
        }
         
        // zero out the arrays, before getting their values for the next iteration.
        for (j=0; j<numADCnames; j++)
        {
          for (k=0; k<maxNumberOfTimeStateVals; k++)
	  {
            timeArray[j][k]=0.0;
            stateArray[j][k]=0;
	  }
        }
        
      }

      xyce_close(p);

      // pointer clean-up
      free(p);
 
      fclose(fptr);

      for (i = 0; i < numADCnames; i++)
      {
        free( ADCnames[i] );
        free( stateArray[i] );
        free( timeArray[i] );
      }
      free( ADCnames );
      free( stateArray );
      free( timeArray );

      vpi_printf("Exiting calltf for ADCStateTest\n");
      return 0;
}

void ADCStateTest_register()
{
      s_vpi_systf_data tf_data;

      tf_data.type      = vpiSysTask;
      tf_data.tfname    = "$ADCStateTest";
      tf_data.calltf    = ADCStateTest_calltf;
      tf_data.compiletf = ADCStateTest_compiletf;
      tf_data.sizetf    = 0;
      tf_data.user_data = 0;
      vpi_register_systf(&tf_data);
}

void (*vlog_startup_routines[])() = {
    ADCStateTest_register,
    0 /* final entry must be zero */
};
